### PR TITLE
add macro average rouge-n to metrax

### PR DESCRIPTION
**Pull request by jshin1394**
macro-average rouge-n matches keras-nlp and torchmetrics implementation

### CODE REVIEW

File: src/metrax/nlp_metrics.py [hunk 1]
```diff
@@ -285,6 +299,159 @@ def compute(self) -> jax.Array:
...
+            - index 2: macro-averaged f1score (derived from avg_precision and
+            avg_recall)
+          Scores are 0.0 if num_examples is zero.
+    """
+    macro_avg_precision = base.divide_no_nan(
+        self.total_precision, self.num_examples
+    )
+    macro_avg_recall = base.divide_no_nan(self.total_recall, self.num_examples)
+    macro_avg_f1score = base.divide_no_nan(self.total_f1, self.num_examples)
+    return jnp.stack([macro_avg_precision, macro_avg_recall, macro_avg_f1score])
```
Comment (jeffcarp) [inline src/metrax/nlp_metrics.py:452]: Is it common for users to want all three of these as a bundle for the most part? (as opposed to separate precision/recall/f1 metrics)

File: src/metrax/nlp_metrics_test.py [hunk 2]
```diff
@@ -86,7 +86,7 @@ def test_bleu_merge(self):
     ]
     predictions = [
         "He He He eats sweet apple which is a fruit",
-        "I love Silicon Valley it's one of my favourite shows",
+        "I love Silicon Valley it is one of my favourite shows",
```
Comment (jeffcarp) [inline src/metrax/nlp_metrics_test.py:89]: Is there a reason behind the diff here?

File: src/metrax/nlp_metrics_test.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def test_bleu(self):
...
     ]
     keras_metric = keras_nlp.metrics.Bleu()
```
Comment (jeffcarp) [inline src/metrax/nlp_metrics_test.py:67]: Sorry I didn't catch this before - keras_nlp became keras_hub so they should be functionally identical. It looks like this metric exists in keras_hub as well:
https://github.com/keras-team/keras-hub/blob/db7b7a4e6e744ed81a1ee673ecb595f791a45a25/keras_hub/src/metrics/bleu.py#L45
(feel free to leave as followup if too disruptive to this PR)

File: src/metrax/nlp_metrics.py [hunk 4]
```diff
@@ -14,28 +14,42 @@
...
+  """Generates a counter of n-grams from a list of tokens for a specific n.
+
+  Args:
+    segment: list. Text segment from which n-grams will be extracted.
+    order: The order of n-grams.
+
+  Returns:
+    A collections.Counter mapping n-gram tuples to their counts.
+  """
+  return collections.Counter(zip(*[segment[i:] for i in range(order)]))
```
Comment (jeffcarp) [inline src/metrax/nlp_metrics.py:36]: Does this make the metric non-jittable?

**Review by jeffcarp — COMMENTED** [review]
Thanks Jiwon!

File: src/metrax/nlp_metrics_test.py (hunk 2 above)
Comment (jshin1394) [reply src/metrax/nlp_metrics_test.py:89]: because of the tokenizer used by us(simple split) vs keras tokenizer, we were seeing different rouge-2 values because of ' processing

File: src/metrax/nlp_metrics_test.py (hunk 3 above)
Comment (jshin1394) [reply src/metrax/nlp_metrics_test.py:67]: will follow up with a PR which modifies the test to use keras_hub.bleu

File: src/metrax/nlp_metrics.py (hunk 4 above)
Comment (jshin1394) [reply src/metrax/nlp_metrics.py:36]: In addition to this line(n-gram generation), other python string operations used within RougeN make the metric non-jittable. Such as taking in list[str] as input and calling pred_str.split().

**Review by jeffcarp — APPROVED** [review]

File: src/metrax/nlp_metrics.py (hunk 1 above)
Comment (jeffcarp) [reply src/metrax/nlp_metrics.py:452]: I meant more would it make sense to just have 3 separate metrics for these, or is it common enough to want all of them that it makes sense to bundle them together? There is likely a reason Keras did it that way but would prefer not to copy over design decisions without understanding first

Comment (jeffcarp) [reply src/metrax/nlp_metrics.py:452]: Looks like the decision originated here:
https://github.com/keras-team/keras-hub/pull/122#discussion_r882341574

In an ideal world I think it would be cleaner if these were separate metric classes. But realistically SG to move forward with this way and make a bug to revisit the decision.
